### PR TITLE
[Feature] Add NPU async DP framework and Async Cam connector

### PR DESCRIPTION
**Pull request by jiangkuaixue123**
## Summary

- Add the NPU async DP framework plumbing, async CAM connector support, and related runtime/config integration.
- Add async DP patch helpers, offline scheduling utilities, PCP debug helpers, and DeepSeek v2 async MoE support.
- Expand unit coverage for async DP patches, connector metadata, offline schedules, config validation, and NPU worker/runtime behavior.

## Impact

This enables the NPU async DP path to be exercised behind the AFD async configuration while keeping the changes isolated in the async connector/runtime flow.

## Validation

- Not run in this environment: local python is unavailable and python3 does not have pytest installed.
- tests/unit/v1/worker/test_npu_runtime.py conflict was resolved and staged before the branch was finalized.

### CODE REVIEW

File: afd_plugin/compat/ascend/ops.py [hunk 1]
```diff
@@ -54,40 +59,85 @@ def _assert_afd_namespace_registered(torch: object) -> None:
     _ = torch.ops.afd_ascend.e2a
 
 
+def _assert_cam_namespace_registered(torch: object) -> None:
+    _dispatch_send = torch.ops.umdk_cam_op_lib.async_dispatch_send
+    _dispatch_recv = torch.ops.umdk_cam_op_lib.async_dispatch_recv
+    _combine_send = torch.ops.umdk_cam_op_lib.async_combine_send
+    _combine_recv = torch.ops.umdk_cam_op_lib.async_combine_recv
+    del _dispatch_send, _dispatch_recv, _combine_send, _combine_recv
```
Comment (jiangkuaixue123) [inline afd_plugin/compat/ascend/ops.py]: 去掉这个del

File: afd_plugin/compat/ascend/runtime.py [hunk 2]
```diff
@@ -88,7 +101,85 @@ def fail_if_unsupported_npu_afd_features(vllm_config: object) -> None:
...
+
+    quant_mode = extra.get("dynamicQuant", extra.get("quant_mode", 0))
+    if quant_mode not in (None, "", 0, "0", 1, "1"):
+        raise RuntimeError(
+            "AFDAsyncConnector currently supports only quant_mode/dynamicQuant 0 or 1",
+        )
+
+
+def _fail_if_unsupported_npu_async_moe_ubatching_features(
+    vllm_config: object,
```
Comment (jiangkuaixue123) [inline afd_plugin/compat/ascend/runtime.py]: 这个文件vllm_config使用vllm的VllmConfig

File: afd_plugin/compat/ascend/runtime.py [hunk 3]
```diff
@@ -88,7 +101,85 @@ def fail_if_unsupported_npu_afd_features(vllm_config: object) -> None:
...
+            f"got async_moe_split={split!r}",
+        )
+    if int(parallel_config.decode_context_parallel_size) > 1:
+        raise RuntimeError(
+            "async_moe_ubatching does not support decode context parallel "
+            "metadata yet",
+        )
+
+
+def _npu_aclgraph_mode_name(vllm_config: object) -> str:
```
Comment (jiangkuaixue123) [inline afd_plugin/compat/ascend/runtime.py]: 不要这个helper函数

File: afd_plugin/connectors/npu/async_cam.py [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+            self.expert_per_rank,
+            self.world_rank,
+            self.topology.world_size,
+            data.layer_idx,
+            self.tp_size,
+            self.dynamic_quant,
+            self.group_name,
+        )
+
+    def recv_ffn_output(self, handle: object = None, **kwargs: object) -> Tensor:
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: 这里的 fallback 可以再简化一下：当前只要 `metadata/topk_ids/topk_weights` 任一缺失，就会先走 `_pop_attention_payload()`；如果没有 pending payload 已经会 raise。所以下面的 `if metadata is None: create_recv_metadata(...)` 分支在当前控制流下基本不可达，可以删掉或把逻辑重排成更明确的两种入口。

File: afd_plugin/connectors/npu/async_cam.py [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+
+
+def _ensure_connector_data(metadata: AFDConnectorMetadata) -> AFDAsyncConnectorData:
+    data = metadata.connector_data
+    if not isinstance(data, AFDAsyncConnectorData):
+        raise RuntimeError("AFD async metadata is missing connector_data")
+    return data
+
+
+def _validate_topk_payload(
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py:892]: 这里的类型可以收窄成 `Tensor` / `Tensor | None`。调用点前面已经要求 `topk_ids/topk_weights` 存在，使用 `object + getattr(shape)` 的防御性写法有点重，直接访问 `.shape` 会更清楚。

File: afd_plugin/connectors/npu/async_cam.py [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+        return
+    formatted_args = "\n".join(
+        f"  {name}={_describe_cam_op_arg(name, value)}"
+        for name, value in kwargs.items()
+        if name not in _CAM_LOG_SKIPPED_ARGS
+    )
+    logger.warning("AFD CAM %s %s:\n%s", op_name, label, formatted_args)
+
+
+def _log_cam_op_inputs(op_name: str, **kwargs: object) -> None:
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: `_log_cam_op_inputs/_log_cam_op_outputs` 只是给 `_log_cam_op_values` 固定 label，封装收益不太大。可以考虑直接调用 `_log_cam_op_values(op_name, "inputs"/"outputs", ...)`，少两个薄 helper。

File: afd_plugin/model_executor/models/deepseek_v2.py [hunk 7]
```diff
@@ -427,14 +753,281 @@ def forward_with_afd(
...
+                ubatch_slice.token_slice,
+                num_tokens=int(hidden_states.shape[0]),
+            )
+            for ubatch_slice in ubatch_slices
+        ]
+
+        moe_start_layer = max(self.start_layer, first_moe_layer)
+        moe_layers = list(islice(self.layers, moe_start_layer, self.end_layer))
+
+        def compute_stage_attention(
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/deepseek_v2.py]: 这几个 nested helper 里的 `layer: Any` 可以标成 `AFDDeepseekV2DecoderLayer`。`moe_layers` 来自 `self.layers` 的 MoE decoder layer 子集，后面也依赖 `layer.layer_idx` 和 `layer.compute_attn_output()`，类型是可以确定的。

File: afd_plugin/model_executor/models/deepseek_v2.py [hunk 8]
```diff
@@ -427,14 +753,281 @@ def forward_with_afd(
...
+            )
+            afd_connector.send_attn_output(
+                stage_hidden_states[stage_idx],
+                stage_metadata,
+                topk_weights=topk_weights,
+                topk_ids=topk_ids,
+                router_logits=router_logits,
+            )
+
+        def recv_stage_ffn(layer: Any, stage_idx: int, event_prefix: str) -> None:
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/deepseek_v2.py]: `recv_stage_ffn` 的 `layer` 和 `event_prefix` 当前没有使用，建议直接删掉这两个参数，调用点也会更轻一点。

File: afd_plugin/v1/worker/ascend/ffn_model_runner.py [hunk 9]
```diff
@@ -403,6 +512,132 @@ def _normalize_recv_output(
...
+    if ffn_output.shared_output is not None:
+        kwargs["expand_x_shared"] = ffn_output.shared_output
+    connector.send_ffn_output(
+        ffn_output.routed_output,
+        metadata,
+        **kwargs,
+    )
+
+
+def _resolve_num_hidden_layers(model_config: object) -> int:
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/ffn_model_runner.py]: 这个 helper 没看到调用点，可以删除。

File: afd_plugin/v1/worker/ascend/ffn_model_runner.py [hunk 10]
```diff
@@ -403,6 +512,132 @@ def _normalize_recv_output(
...
+def _is_moe_layer(hf_config: object, layer_idx: int) -> bool:
+    moe_layer_freq = getattr(hf_config, "moe_layer_freq", 1)
+    return (
+        hf_config.n_routed_experts is not None
+        and layer_idx >= hf_config.first_k_dense_replace
+        and layer_idx % moe_layer_freq == 0
+    )
+
+
+def _first_dp_token_counts(dp_metadata_list: dict[int, Any]) -> Any:
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/ffn_model_runner.py]: 这个 helper 也没有调用点，可以删除。

File: tests/unit/test_envs.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+    assert force_balanced_topk_ids_enabled() is False
+
+
+def test_force_balanced_topk_ids_env_accepts_true_values(monkeypatch):
+    for value in ("1", "true", "yes", "on"):
+        monkeypatch.setenv(AFD_FORCE_BALANCED_TOPK_IDS, value)
+
+        assert force_balanced_topk_ids_enabled() is True
+
```
Comment (jiangkuaixue123) [inline tests/unit/test_envs.py]: `git diff --check` 报这里末尾有多余空行，合并前顺手清一下。

File: afd_plugin/model_executor/models/deepseek_v2.py [hunk 12]
```diff
@@ -427,14 +753,281 @@ def forward_with_afd(
         )
         return hidden_states, residual
 
+    def forward_with_afd_v2(
+        self,
+        hidden_states: torch.Tensor,
+        residual: torch.Tensor | None,
+        positions: torch.Tensor,
+        afd_metadata: object,
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/deepseek_v2.py]: 这里以及 `forward_with_afd` / `forward_with_afd_v3` 的 `afd_metadata: object` 可以明确标成 `AFDMetadata`。调用方来自 `get_afd_metadata_from_forward_context()`，后面也直接访问 `.afd_connector/.afd_stage_idx/.clone()` 等 AFDMetadata 字段，类型收窄后能减少读代码时的猜测。

File: afd_plugin/connectors/npu/async_cam.py [hunk 13]
```diff
@@ -0,0 +1,746 @@
...
+
+    @property
+    def world_size(self) -> int:
+        return self.attention_rank_size + self.expert_rank_size
+
+
+class AFDAsyncConnector(AFDConnectorBase):
+    """CAM-backed async-DP connector for Ascend NPU AFD.
+
+    Phase 1 owns only the connector contract, CAM op call shape, and metadata
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: 这个pr去掉类似Phase等开发过程阶段的描述

File: afd_plugin/connectors/npu/async_cam.py [hunk 14]
```diff
@@ -0,0 +1,746 @@
...
+    uses_dp_metadata_control_plane = False
+    ffn_step_trigger = "connector"
+    requires_eager = True
+    required_platform = "ascend"
+
+    def __init__(
+        self,
+        rank: int,
+        local_rank: int,
+        vllm_config: object,
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: 具体类型VllmConfig

File: afd_plugin/connectors/npu/async_cam.py [hunk 15]
```diff
@@ -0,0 +1,746 @@
...
+            )
+            or "",
+        )
+        self.max_seq_len = max(
+            1,
+            int(vllm_config.scheduler_config.max_num_batched_tokens),
+        )
+        self.comm_id = CAM_COMM_ID
+        self.tp_size = _resolve_cam_tp_size(afd_config)
+        self.cam_pg: Any | None = None
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: self.cam_pg的类型

File: afd_plugin/model_executor/models/forward_context.py [hunk 16]
```diff
@@ -31,6 +33,20 @@ def get_afd_metadata_from_forward_context(forward_context: object | None = None)
     return getattr(forward_context, "afd_metadata", None)
 
 
+def get_async_moe_ubatch_metadata_from_forward_context(
+    forward_context: object | None = None,
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/forward_context.py]: forward_context应该是有具体类型的

File: afd_plugin/v1/worker/ascend/attention_model_runner.py [hunk 17]
```diff
@@ -191,6 +218,67 @@ def _build_attention_metadata(self, *args: Any, **kwargs: Any) -> Any:
...
+            return full_metadata
+
+        ubatch_slices = create_request_boundary_ubatch_slices(
+            num_scheduled_tokens_np,
+            num_ubatches=async_moe_num_ubatches(self.afd_config),
+        )
+        if ubatch_slices is None:
+            return full_metadata
+
+        logger.warning(
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/attention_model_runner.py]: 这个warning改成debug

File: afd_plugin/v1/worker/ascend/attention_model_runner.py [hunk 18]
```diff
@@ -243,6 +331,124 @@ def _get_pcp_metadata(block_table_tensor):
                 num_reqs,
             )
 
+        def _build_stage_local_pcp_metadata(
+            common_attn_metadata: Any,
+            ubatch_slice: Any,
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/attention_model_runner.py]: 上面两个any应该是有具体类型的

File: afd_plugin/v1/worker/ascend/attention_model_runner.py [hunk 19]
```diff
@@ -1600,7 +1771,17 @@ def _model_forward_values(
...
+    "logits_indices",
+    "use_spec_decode",
+    "for_cudagraph_capture",
+    "num_scheduled_tokens",
+    "num_scheduled_tokens_np",
+    "cascade_attn_prefix_lens",
 ]
 
 
+def _is_npu_ubatching_enabled(vllm_config: object) -> bool:
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/attention_model_runner.py]: 这个helper函数请内连掉

File: afd_plugin/v1/worker/ascend/attention_model_runner.py [hunk 20]
```diff
@@ -1024,6 +1246,18 @@ def _install_afd_metadata_on_forward_context(
             dp_metadata = self._build_capture_dp_metadata(padded_graph_tokens)
         self._send_dp_metadata(dp_metadata, ubatch_slices)
 
+    def _install_async_moe_ubatch_metadata_on_forward_context(
+        self,
+        forward_context: object,
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/attention_model_runner.py]: forward_context的具体类型

File: afd_plugin/model_executor/models/forward_context.py [hunk 21]
```diff
@@ -31,6 +35,20 @@ def get_afd_metadata_from_forward_context(forward_context: object | None = None)
     return getattr(forward_context, "afd_metadata", None)
 
 
+def get_async_moe_ubatch_metadata_from_forward_context(
+    forward_context: ForwardContext | None = None,
+) -> Any:
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/forward_context.py]: 这几个return的Any 推导一下类型

File: afd_plugin/config.py [hunk 22]
```diff
@@ -224,6 +287,27 @@ def parse_afd_config(
...
+def is_afd_async_dp(vllm_config: VllmConfig) -> bool:
+    """Return whether ``vllm_config`` selects AFD's async connector mode."""
+
+    config = parse_afd_config(vllm_config, validate=False)
+    return (
+        config.enabled and config.async_dp and config.connector == AFD_ASYNC_CONNECTOR
+    )
+
+
+def is_afd_async_attention_dp(vllm_config: VllmConfig) -> bool:
```
Comment (jiangkuaixue123) [inline afd_plugin/config.py]: 有了is_afd_async_dp 不需要is_afd_async_attention_dp了

File: afd_plugin/v1/worker/ascend/ffn_worker.py [hunk 23]
```diff
@@ -157,4 +160,16 @@ def shutdown(self) -> None:
         super().shutdown()
 
 
+def _set_npu_device_if_possible(device: object) -> None:
+    if device.type != "npu":
+        return
+    torch.npu.set_device(device)
+
+
+def _synchronize_npu_if_possible(device: object) -> None:
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/ffn_worker.py]: _synchronize_npu_if_possible _set_npu_device_if_possible这两个过渡防御了 跑进这儿的就是npu的 不需要这两个helper

File: afd_plugin/model_executor/models/forward_context.py [hunk 24]
```diff
@@ -7,13 +7,25 @@
...
+from vllm.v1.worker.ubatch_utils import UBatchSlices
 
+from afd_plugin.connectors import AFDMetadata
 
-def get_afd_metadata_from_forward_context(forward_context: object | None = None) -> Any:
+ASYNC_MOE_UBATCH_METADATA_KEY: Final[str] = "afd_async_moe_ubatch_metadata"
+
+
+class AsyncMoeUbatchMetadata(TypedDict):
+    attn_metadata: object
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/forward_context.py:22]: 这个的具体类型

File: afd_plugin/connectors/npu/async_cam.py [hunk 25]
```diff
@@ -0,0 +1,960 @@
...
+if TYPE_CHECKING:
+    from torch.distributed.distributed_c10d import ProcessGroup
+    from vllm.config import VllmConfig
+
+AFD_ASYNC_CAM_GROUP_NAME = "afd_async_cam"
+CAM_COMM_ID = 0
+ATTN_RANKS_PER_DP_CONFIG_KEY = "attn_ranks_per_dp"
+
+try:
+    from vllm.logger import init_logger
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: 不用try catch 就用from vllm.logger import init_logger

File: afd_plugin/connectors/npu/async_cam.py [hunk 26]
```diff
@@ -0,0 +1,960 @@
...
+    if ffn_output.shared_output is not None:
+        kwargs["expand_x_shared"] = ffn_output.shared_output
+    connector.send_ffn_output(
+        ffn_output.routed_output,
+        metadata,
+        **kwargs,
+    )
+
+
+def _ensure_connector_data(metadata: AFDConnectorMetadata) -> AFDAsyncConnectorData:
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: 去掉这个_ensure 我不需要

File: afd_plugin/v1/worker/ascend/ffn_model_runner.py [hunk 27]
```diff
@@ -513,6 +627,13 @@ def _to_dp_level_token_counts(
     return num_tokens_across_dp[indices].contiguous()
 
 
+def _resolve_graph_pool() -> object | None:
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/ascend/ffn_model_runner.py]: 去掉这个helper函数
